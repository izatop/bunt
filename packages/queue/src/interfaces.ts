import {IContext} from "@typesafeunit/unit";
import {Promisify} from "@typesafeunit/util";
import {Handler} from "./Handler";
import {Message, MessagePayload, TaskAbstract} from "./Queue";

export type ActionHandler<C extends IContext, M extends Message> = Handler<C, M>;

export type HandlerReturn<M extends Message> = M extends TaskAbstract<any, infer R>
    ? Promisify<R>
    : Promisify<any>;

export type HandlerState<M extends Message> = { payload: MessagePayload<M> };
