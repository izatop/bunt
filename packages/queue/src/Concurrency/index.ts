export * from "./ConcurrencyQueue";
export * from "./RateLimitedQueue";
