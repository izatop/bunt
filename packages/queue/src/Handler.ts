import {Action, Context} from "@bunt/unit";
import {Promisify} from "@bunt/util";
import {HandlerReturn, HandlerState} from "./interfaces.js";
import {Message, MessagePayload} from "./Queue/index.js";

export abstract class Handler<C extends Context, M extends Message>
    extends Action<C, HandlerState<M>, HandlerReturn<M>> {
    protected get payload(): MessagePayload<M> {
        return this.state.payload;
    }

    public abstract run(): Promisify<HandlerReturn<M>>;
}
