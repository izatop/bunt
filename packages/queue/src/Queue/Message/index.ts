export * from "./MessageAbstract";
export * from "./TransactionAbstract";
export * from "./TaskAbstract";
