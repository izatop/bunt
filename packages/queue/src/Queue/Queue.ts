import {ITransport} from "../interfaces";
import {QueueAbstract} from "./QueueAbstract";

export class Queue<Q extends ITransport> extends QueueAbstract<Q> {

}
