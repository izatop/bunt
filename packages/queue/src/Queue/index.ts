export * from "./fn";
export * from "./interfaces";
export * from "./QueueAbstract";
export * from "./ReadOperation";
export * from "./SubscriptionAbstract";
export * from "./Queue";
export * from "./Message";
