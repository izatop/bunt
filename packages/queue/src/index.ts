export * from "./Queue";
export * from "./Redis";
export * from "./Handler";
export * from "./interfaces";
export * from "./Dispatcher";
export * from "./PubSub";
