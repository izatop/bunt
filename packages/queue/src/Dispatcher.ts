import {
    ActionFactory,
    ActionTransactionHandlers,
    Context,
    ContextArg,
    Disposer,
    Heartbeat,
    IRunnable,
    Unit,
    unit,
} from "@bunt/unit";
import {Ctor, Defer, logger, Logger} from "@bunt/util";
import {Handler} from "./Handler";
import {ITransport} from "./interfaces";
import {Incoming, MessageCtor, Queue, QueueAbstract} from "./Queue";

export class Dispatcher<C extends Context> extends Disposer implements IRunnable {
    @logger
    public logger!: Logger;

    readonly #unit: Unit<C>;
    readonly #queue: QueueAbstract<ITransport>;
    readonly #route = new Map<MessageCtor<any>, ActionFactory<any>>();

    protected constructor(u: Unit<C>, queue: QueueAbstract<ITransport>) {
        super();

        this.#unit = u;
        this.#queue = queue;
        this.onDispose(queue);
    }

    public get size(): number {
        return this.#route.size;
    }

    public static async factory<C extends Context>(
        context: ContextArg<C>, queue: Queue<ITransport>): Promise<Dispatcher<C>> {

        return new this(await unit(context), queue);
    }

    public getHeartbeat(): Heartbeat {
        const running = new Defer<void>();
        this.onDispose(() => running.resolve());

        return Heartbeat.create(this)
            .enqueue(running)
            .onDispose(this);
    }

    public subscribe<M extends Incoming, H extends Handler<C, M>>(type: MessageCtor<M>, action: Ctor<H>): this {
        const subscription = this.#queue.on<any>(type, ({payload}) => this.#unit.run(action, {payload}));

        this.onDispose(subscription);

        return this;
    }

    public on(handlers: ActionTransactionHandlers) {
        this.#unit.on(handlers);
    }
}
