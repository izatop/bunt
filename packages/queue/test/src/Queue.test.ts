import {dispose} from "@bunt/unit";
import {wait} from "@bunt/util";
import {Queue} from "../../src";
import {BarMessage} from "./Message/BarMessage";
import {FooMessage} from "./Message/FooMessage";
import {MultiplyTask} from "./Message/MultiplyTask";
import {NumMessage} from "./Message/NumMessage";
import {TestTransport} from "./Queue/TestTransport";

describe("Queue", () => {
    test("Subscription", async () => {
        const success: FooMessage[] = [];
        const queue = new Queue(new TestTransport());
        const sub1 = queue.on(FooMessage, (message) => success.push(message));

        expect(sub1.subscribed).toBe(true);
        await expect(sub1.subscribe()).rejects.toThrow();
        await expect(sub1.unsubscribe()).resolves.toBeUndefined();

        queue.send(new FooMessage(true));
        await sub1.subscribe();

        expect(success).toMatchSnapshot();

        await dispose(queue);
    });

    test("Rejection", async () => {
        const rejections: any[] = [];
        const queue = new Queue(new TestTransport());

        queue.send(new BarMessage(1));
        const sub = queue.on(BarMessage, () => {
            throw new Error("Test Error");
        });

        sub.watch(({error, status, message}) => {
            rejections.push({error, status, message});
        });

        await sub.unsubscribe();
        expect(rejections.length).toBe(1);
        expect(rejections[0].status).toBe(false);
        expect(rejections).toMatchSnapshot();

        await dispose(queue);
    });

    test("Task", async () => {
        const queue = new Queue(new TestTransport());
        const calc = [10, 5, 3];
        queue.send(new MultiplyTask(calc));
        queue.on(MultiplyTask, ({payload}) => {
            return payload.reduce((l, r) => l + r, 0);
        });

        const reply = await wait<NumMessage>((resolve) => {
            queue.on(NumMessage, (message) => resolve(message));
        });

        expect(reply.payload).toBe(calc.reduce((l, r) => l + r, 0));

        await dispose(queue);
    });
});
