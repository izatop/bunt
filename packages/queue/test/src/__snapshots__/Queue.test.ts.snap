// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queue Rejection 1`] = `
Array [
  Object {
    "error": [Error: Test Error],
    "message": BarMessage {
      "payload": 1,
    },
    "status": false,
  },
]
`;

exports[`Queue Subscription 1`] = `
Array [
  FooMessage {
    "payload": true,
  },
]
`;
